compliance/virtualmachines/relay/relay.go

```diff
@@ -81,38 +115,56 @@ func (r *Relay) Run() error {
 	}()
 
 	for {
+		if err := r.vsockServer.acquireSemaphore(r.ctx); err != nil {
+			if r.ctx.Err() != nil {
+				log.Info("Stopping virtual machine relay")
+				return r.ctx.Err()
+			}
+
+			log.Warnf("Failed to acquire semaphore to handle connection after %v: %v",
+				r.vsockServer.maxSemaphoreWaitTime, err,
+			)
+
+			continue
+		}
+
 		// Accept() is blocking, but it will return when ctx is cancelled and the above goroutine calls r.vsockServer.Stop()
 		conn, err := r.vsockServer.listener.Accept()
 		if err != nil {
+			r.vsockServer.releaseSemaphore()
+
 			if r.ctx.Err() != nil {
 				log.Info("Stopping virtual machine relay")
 				return r.ctx.Err()
 			}
-			// We deliberately not kill the listener on errors. The only way to stop that is to cancel the context.
+
+			// We deliberately don't kill the listener on errors. The only way to stop that is to cancel the context.
 			// If we had return here on fatal errors, then compliance would continue working without the relay
 			// and that would make it an invisible problem to the user.
 			log.Errorf("Error accepting connection: %v", err)
+
 			time.Sleep(r.waitAfterFailedAccept) // Prevent a tight loop
 			continue
 		}
+		metrics.VsockConnectionsAccepted.Inc()
 
 		go func(conn net.Conn) {
+			defer r.vsockServer.releaseSemaphore()
+
 			defer func(conn net.Conn) {
 				if err := conn.Close(); err != nil {
 					log.Errorf("Failed to close connection: %v", err)
 				}
 			}(conn)
 
-			if err := r.handleVsockConnection(conn); err != nil {
-				log.Errorf("Error handling vsock connection: %v", err)
+			if err = r.handleVsockConnection(conn); err != nil {
```

Comment:
nit: I think here it does not hurt to shadow the err variable from the Run function. It would make sure that the variable is consumed in the scope where it is created (which is the case with the log).